File: src/main/java/cilly/FieldInfo.java
/*
 * System.Reflection-like API for access to .NET assemblies (DLL & EXE)
 */

package cilly;

/**
 * Discovers the attributes of a field and provides access to field metadata.
 * 
 * @author Nikolay Mihaylov
 * @version 1.0
 */
public class FieldInfo extends MemberInfo implements HasCustomModifiers {

    // ##########################################################################
    // public interface

    @Override
    public final int memberType() {
        return MemberTypes.Field;
    }

    /** Attributes associated with this field. */
    public final short attributes;

    /** Type of the field represented by this FieldInfo object. */
    public final Type fieldType;

    /** can be null */
    public final CustomModifier[] cmods;

    protected final Object value;

    public final boolean isStatic() {
        return (attributes & FieldAttributes.Static) != 0;
    }

    public final boolean isInitOnly() {
        return (attributes & FieldAttributes.InitOnly) != 0;
    }

    public final boolean isLiteral() {
        return (attributes & FieldAttributes.Literal) != 0;

    }

    public final boolean isPublic() {
        return (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Public;
    }

    public final boolean isPrivate() {
        return (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Private;
    }

    public final boolean isFamily() {
        return (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Family;
    }

    public final boolean isAssembly() {
        return (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Assembly;
    }

    public final boolean isFamilyOrAssembly() {
        return (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.FamORAssem;
    }

    public final boolean isFamilyAndAssembly() {
        return (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.FamANDAssem;
    }

    public final boolean isSpecialName() {
        return (attributes & FieldAttributes.SpecialName) != 0;
    }

    public final boolean isPinvokeImpl() {
        return (attributes & FieldAttributes.PinvokeImpl) != 0;
    }

    public final boolean isNotSerialized() {
        return (attributes & FieldAttributes.NotSerialized) != 0;
    }

    private boolean knownVolatile = false;
    private boolean cachedVolatile = false;

    public final boolean isVolatile() {
        if (knownVolatile)
            return cachedVolatile;
        knownVolatile = true;
        if (cmods == null) {
            cachedVolatile = false;
            return cachedVolatile;
        }
        for (int idx = 0; idx < cmods.length; idx++) {
            if (cmods[idx].marker == CustomModifier.VolatileMarker()) {
                cachedVolatile = true;
                return cachedVolatile;
            }
        }
        cachedVolatile = false;
        return cachedVolatile;
    }

    @Override
    public final Type[] getOptionalCustomModifiers() {
        return CustomModifier.helperCustomMods(false, cmods);
    }

    @Override
    public final Type[] getRequiredCustomModifiers() {
        return CustomModifier.helperCustomMods(true, cmods);
    }

    @Override
    public String toString() {
        return FieldAttributes.toString(attributes) + " " + fieldType + " "
                + declaringType.fullName + "::" + name;
    }

    // ##########################################################################

    protected static final FieldInfo[] EMPTY_ARRAY = new FieldInfo[0];

    /** Initializes a new instance of the FieldInfo class. */
    protected FieldInfo(String name, Type declType, int attrs,
            cilly.util.PECustomMod fieldTypeWithMods, Object value) {
        super(name, declType);
        fieldType = fieldTypeWithMods.marked();
        cmods = fieldTypeWithMods.cmods();
        attributes = (short) attrs;
        this.value = value;
        this.name = name;
        this.declaringType = declType;
    }
    
    public final String name;
    public final Type declaringType;

    /**
     */
    public Object getValue() {
        return value;
    }

    // ##########################################################################

} // class FieldInfo

File: src/main/scala/cilly/FieldInfo.scala
/*
 * System.Reflection-like API for access to .NET assemblies (DLL & EXE)
 */
package cilly

import cilly.util.PECustomMod

/**
 * Discovers the attributes of a field and provides access to field metadata.
 *
 * @author Nikolay Mihaylov
 * @version 1.0
 */
class FieldInfo(override val name: String, override val declaringType: Type, val attributes: Short, val fieldTypeWithMods: PECustomMod, val value: AnyRef) extends MemberInfo(name, declaringType) with HasCustomModifiers {
  /** Type of the field represented by this FieldInfo object. */
  val fieldType: Type = fieldTypeWithMods.marked
  val cmods: Array[CustomModifier] = fieldTypeWithMods.cmods

  final def memberType: Int = MemberTypes.Field

  final def isStatic: Boolean = (attributes & FieldAttributes.Static) != 0

  final def isInitOnly: Boolean = (attributes & FieldAttributes.InitOnly) != 0

  final def isLiteral: Boolean = (attributes & FieldAttributes.Literal) != 0

  final def isPublic: Boolean = (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Public

  final def isPrivate: Boolean = (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Private

  final def isFamily: Boolean = (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Family

  final def isAssembly: Boolean = (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.Assembly

  final def isFamilyOrAssembly: Boolean = (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.FamORAssem

  final def isFamilyAndAssembly: Boolean = (attributes & FieldAttributes.FieldAccessMask) == FieldAttributes.FamANDAssem

  final def isSpecialName: Boolean = (attributes & FieldAttributes.SpecialName) != 0

  final def isPinvokeImpl: Boolean = (attributes & FieldAttributes.PinvokeImpl) != 0

  final def isNotSerialized: Boolean = (attributes & FieldAttributes.NotSerialized) != 0

  private var knownVolatile: Boolean = false
  private var cachedVolatile: Boolean = false

  final def isVolatile: Boolean = {
    if (knownVolatile) return cachedVolatile
    knownVolatile = true
    if (cmods == null) {
      cachedVolatile = false
      return cachedVolatile
    }
    {
      var idx: Int = 0
      while (idx < cmods.length) {
        {
          if (cmods(idx).marker eq CustomModifier.VolatileMarker) {
            cachedVolatile = true
            return cachedVolatile
          }
        }
        ({
          idx += 1; idx - 1
        })
      }
    }
    cachedVolatile = false
    cachedVolatile
  }

  final def getOptionalCustomModifiers: Array[Type] = CustomModifier.helperCustomMods(false, cmods)

  final def getRequiredCustomModifiers: Array[Type] = CustomModifier.helperCustomMods(true, cmods)

  override def toString: String = FieldAttributes.toString(attributes) + " " + fieldType + " " + declaringType.fullName + "::" + name

  /**
    */
  def getValue: AnyRef = value
}